File: Maximum Twin Sum of a Linked List.cpp
class Solution {
public:
    int pairSum(ListNode* head) {
        int max = 0;
        ListNode *sgl = head;
        ListNode *dbl = head;
        while(dbl!=NULL){
            dbl = dbl->next->next;
            sgl = sgl->next;
        }
        ListNode *secondhalf = reverseList(sgl);
        ListNode *second = secondhalf;
        dbl = head;
        while(second != NULL){
            int sum = dbl->val + second->val;
            if(sum > max)
                max = sum;
            dbl = dbl->next;
            second = second->next;
        }
        return max;
    }
    ListNode* reverseList(ListNode *h){
        if(h == NULL)
            return 0;
        if(!h->next || !h)
            return h;
        auto answer = reverseList(h->next);
        h->next->next = h;
        h->next = NULL;
        return answer;
        }
};

File: Maximum Twin Sum of a Linked List.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def pairSum(self, head: Optional[ListNode]) -> int:
        size = 0
        temp = head
        while temp:
            temp = temp.next
            size += 1

        curr = head
        for i in range(size//2):
            curr = curr.next
        nodes = []
        temp = curr
        half = curr
        while temp:
            nodes.append(temp.val)
            temp = temp.next
        nodes.reverse()
        for node in nodes:
            curr.val = node
            curr = curr.next
        maxsum = 0
        left = head
        while half:
            maxsum = max(maxsum, left.val+half.val)
            left = left.next
            half = half.next
        return maxsum
